operators/src/processing/new_expression/query_processor.rs

```diff
@@ -0,0 +1,401 @@
+            .map(|(a, b)| {
+                let is_a_no_data = tile_0.is_no_data(*a);
+                let is_b_no_data = tile_1.is_no_data(*b);
+
+                if !map_no_data && (is_a_no_data || is_b_no_data) {
+                    return out_no_data;
+                }
+
+                let result = expression(
+                    a.as_(),
+                    is_a_no_data,
+                    b.as_(),
+                    is_b_no_data,
+                    out_no_data.as_(),
+                );
+                TO::from_(result)
+            })
+            .collect();
+
+        Result::<Vec<TO>>::Ok(data)
+    }
+}
+
+#[async_trait]
+impl<TO, T1, T2, T3> ExpressionTupleProcessor<TO>
+    for (
+        BoxRasterQueryProcessor<T1>,
+        BoxRasterQueryProcessor<T2>,
+        BoxRasterQueryProcessor<T3>,
+    )
+where
+    TO: Pixel,
+    T1: Pixel + AsPrimitive<TO>,
+    T2: Pixel,
+    T3: Pixel,
+{
+    type Tuple = (RasterTile2D<T1>, RasterTile2D<T2>, RasterTile2D<T3>);
+
+    #[inline]
+    async fn queries<'a>(
+        &'a self,
+        query: RasterQueryRectangle,
+        ctx: &'a dyn QueryContext,
+    ) -> Result<BoxStream<'a, Result<Self::Tuple>>> {
+        // TODO: tile alignment
+
+        let queries = try_join!(
+            self.0.query(query, ctx),
+            self.1.query(query, ctx),
+            self.2.query(query, ctx)
+        )?;
+
+        let stream =
+            StreamTupleZip::new(queries).map(|rasters| Ok((rasters.0?, rasters.1?, rasters.2?)));
+
+        Ok(stream.boxed())
+    }
+
+    #[inline]
+    fn all_empty(tuple: &Self::Tuple) -> bool {
+        tuple.0.grid_array.is_empty()
+            && tuple.1.grid_array.is_empty()
+            && tuple.2.grid_array.is_empty()
+    }
+
+    #[inline]
+    fn empty_raster(tuple: &Self::Tuple) -> RasterTile2D<TO> {
+        tuple.0.clone().convert()
+    }
+
+    #[inline]
+    fn metadata(tuple: &Self::Tuple) -> (TimeInterval, GridIdx2D, GeoTransform, GridShape2D) {
+        let raster = &tuple.0;
+
+        (
+            raster.time,
+            raster.tile_position,
+            raster.global_geo_transform,
+            raster.grid_shape(),
+        )
+    }
+
+    fn compute_expression(
+        rasters: Self::Tuple,
+        program: &LinkedExpression,
+        map_no_data: bool,
+        out_no_data: TO,
+    ) -> Result<Vec<TO>> {
+        let expression = unsafe {
+            // we have to "trust" that the function has the signature we expect
+            program.function_7::<f64, bool, f64, bool, f64, bool, f64>()?
+        };
+
+        // TODO: allow iterating over empty rasters
+        let tile_0 = rasters.0.into_materialized_tile();
+        let tile_1 = rasters.1.into_materialized_tile();
+        let tile_2 = rasters.2.into_materialized_tile();
+
+        let data = (
+            &tile_0.grid_array.data,
+            &tile_1.grid_array.data,
+            &tile_2.grid_array.data,
+        )
+            .into_par_iter()
```

Comment:
I think we need to chunk here

Reply:
It is not too high, at least it outperforms OpenCL int the benchmark. And also the experiments prior to the implementation showed that this is not a major problem.

However, in a separate task, one could measure this and try to improve the performance.